database/src/main/java/com/firebase/ui/database/FirebaseRecyclerAdapter.java

```diff
@@ -88,6 +108,17 @@ public void startListening() {
     @Override
     public void cleanup() {
         mSnapshots.removeChangeEventListener(this);
+        notifyDataSetChanged();
+    }
+
+    @SuppressWarnings("unused")
+    @OnLifecycleEvent(Lifecycle.Event.ON_ANY)
+    void cleanup(LifecycleOwner source, Lifecycle.Event event) {
+        if (event == Lifecycle.Event.ON_STOP) {
+            cleanup();
+        } else if (event == Lifecycle.Event.ON_DESTROY) {
```

Comment:
Yeah, I'm 99.9% sure it will work because I copied this from the Android source code for removing `LiveData` observers:
```java
@SuppressWarnings("unused")
@OnLifecycleEvent(Lifecycle.Event.ON_ANY)
void onStateChange() {
    if (owner.getLifecycle().getCurrentState() == DESTROYED) {
        removeObserver(observer);
        return;
    }
    // immediately set active state, so we'd never dispatch anything to inactive
    // owner
    activeStateChanged(isActiveState(owner.getLifecycle().getCurrentState()));
}
```